### PR TITLE
Advent of code - Day 1 -> 15

### PR DESCRIPTION
**Pull request by R4PaSs**
More examples with the problems from [Advent of Code](adventofcode.com), day 1 to day 15.

### CODE REVIEW

File: examples/advent_of_code/day12/jsabascus.nit [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+if args.is_empty then
+	print "Usage: ./jsabascus input.txt"
+	exit 1
+end
+
+var json = args[0].to_path.read_all.parse_json
+
+if json == null then
+	print 0
+	exit 0
```
Comment (privat) [inline examples/advent_of_code/day12/jsabascus.nit]: else print json.sum_values

exit should be prohibited, return is acceptable

File: examples/advent_of_code/day12/jsabascus_nored.nit [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+		if has_red then return 0
+		for k, v in self do
+			if v == null then continue
+			sum += v.sum_values
+		end
+		return sum
+	end
+
+	redef fun has_red do
+		for k, v in self do if v == "red" then return true
```
Comment (privat) [inline examples/advent_of_code/day12/jsabascus_nored.nit]: return values.has("red") # Red is dead

File: examples/advent_of_code/day12/jsabascus_nored.nit [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+redef class JsonArray
+	redef fun sum_values do
+		var sum = 0
+		for i in self do
+			if i == null then continue
+			sum += i.sum_values
+		end
+		return sum
+	end
+end
```
Comment (privat) [inline examples/advent_of_code/day12/jsabascus_nored.nit]: maybe import the part1 to factorize code

Comment (privat) [inline examples/advent_of_code/day12/jsabascus_nored.nit]: you can ever reuse the main part!

File: examples/advent_of_code/day10/elves_look_elves_say.nit [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+				cnt += 1
+			else
+				b.append(cnt.to_s)
+				b.add its[i - 1].ascii
+				cnt = 1
+			end
+		end
+		b.append(cnt.to_s)
+		b.add(chars.last)
+		return b.to_s
```
Comment (privat) [inline examples/advent_of_code/day10/elves_look_elves_say.nit]: all this seems more complex than it should :/

File: examples/advent_of_code/day11/corporate_policy.nit [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	end
+
+	fun pairs: MultiHashMap[Text, Int] do
+		var pairs = new MultiHashMap[Text, Int]
+		for i in [1 .. length[ do
+			if self[i - 1] == self[i] then
+				pairs[substring(i - 1, 2)].add(i - 1)
+			end
+		end
+		return pairs
```
Comment (privat) [inline examples/advent_of_code/day11/corporate_policy.nit]: overkill

File: examples/advent_of_code/day13/knights_dinner_table_2.nit [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	var pers_to = els.last
+	pers_to = pers_to.substring(0, pers_to.length - 1)
+	if not persons.has_key(pers_from) then persons[pers_from] = new Person(pers_from)
+	if not persons.has_key(pers_to) then persons[pers_to] = new Person(pers_to)
+	var from = persons[pers_from]
+	var to = persons[pers_to]
+	from.happiness_map[to] = amount
+end
+
+var me = new Person("Myself")
```
Comment (privat) [inline examples/advent_of_code/day13/knights_dinner_table_2.nit]: please import part 1

File: examples/advent_of_code/day13/knights_dinner_table.nit [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+		if remaining.length == 1 then return [[self]]
+		var persons = new Array[Array[Person]]
+		var rem_set = remaining.clone
+		rem_set.remove self
+		for i in rem_set do
+			persons.add_all i.compute_seatings(rem_set)
+		end
+		for i in persons do i.unshift self
+		return persons
+	end
```
Comment (privat) [inline examples/advent_of_code/day13/knights_dinner_table.nit]: use combinations?

~~~nit
var res = new CombinationCollection[Int](remaining, remaining.length)
res.are_sorted = false
res.are_unique = true
return res
~~~

File: examples/advent_of_code/day13/knights_dinner_table.nit [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+fun happiness(arr: Array[Person]): Int do
+	var curr = arr.first
+	var arrln = arr.length
+	var lft = arr.last
+	var rgt = arr[1]
+	var happiness = curr.happiness_map[lft] + curr.happiness_map[rgt]
+	for i in [1 .. arrln - 1[ do
+		var pers = arr[i]
+		lft = arr[i - 1]
+		rgt = arr[i + 1]
```
Comment (privat) [inline examples/advent_of_code/day13/knights_dinner_table.nit]: use `Sequence::modulo` and remove special border cases

File: examples/advent_of_code/day14/reindeer_olympics.nit [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+	var speed = pts[3].to_i
+	var endur = pts[6].to_i
+	var rest = pts[pts.length - 2].to_i
+	deers.add(new Reindeer(name, speed, endur, rest))
+end
+
+var points = new HashMap[Reindeer, Int]
+for i in deers do points[i] = 0
+
+var curr_dist = new HashMap[Reindeer, Int]
```
Comment (privat) [inline examples/advent_of_code/day14/reindeer_olympics.nit]: shouldn't they be attributes in Reindeer?

**Comment by privat** [comment]
this is nice

File: examples/advent_of_code/day13/knights_dinner_table.nit (hunk 7 above)
Comment (R4PaSs) [inline examples/advent_of_code/day13/knights_dinner_table.nit]: Welp. Didn't work, either I'm using it wrong or there's a bug, in any case I'll leave the actual implementation for now

**Comment by privat** [comment]
move to contrib and pass the tests

**Comment by privat** [comment]
I think that the code is in a better place in https://github.com/R4PaSs/advent_of_code_2016
I can even add it as an oot project

**Pull request closed without merging**